File: head_tracking_audio/python_host/audio/latency_monitor.py
import time
from collections import deque


class LatencyMonitor:
	def __init__(self, window=200):
		self.window = int(window)
		self.samples = deque(maxlen=self.window)
		self.queue_age_samples = deque(maxlen=self.window)
		self.uart_delay_samples = deque(maxlen=self.window)
		self._start = None

	def start_callback(self):
		self._start = time.perf_counter()

	def end_callback(self, frames, sample_rate):
		if self._start is None:
			return

		elapsed = time.perf_counter() - self._start
		budget = float(frames) / float(sample_rate) if sample_rate > 0 else 0.0
		self.samples.append((elapsed, budget))
		self._start = None

	def record_queue_age(self, age_sec):
		if age_sec is None:
			return
		self.queue_age_samples.append(float(age_sec))

	def record_uart_delay(self, delay_sec):
		if delay_sec is None:
			return
		self.uart_delay_samples.append(float(delay_sec))

	def stats(self):
		if not self.samples and not self.queue_age_samples and not self.uart_delay_samples:
			return None

		stats = {}
		if self.samples:
			elapsed = [s[0] for s in self.samples]
			budget = [s[1] for s in self.samples]
			stats["avg_ms"] = 1000.0 * (sum(elapsed) / len(elapsed))
			stats["max_ms"] = 1000.0 * max(elapsed)
			stats["overruns"] = sum(1 for e, b in self.samples if b > 0.0 and e > b)
			stats["budget_ms"] = 1000.0 * (sum(budget) / len(budget)) if budget else 0.0

		if self.queue_age_samples:
			stats["queue_age_avg_ms"] = 1000.0 * (sum(self.queue_age_samples) / len(self.queue_age_samples))
			stats["queue_age_max_ms"] = 1000.0 * max(self.queue_age_samples)

		if self.uart_delay_samples:
			stats["uart_delay_avg_ms"] = 1000.0 * (sum(self.uart_delay_samples) / len(self.uart_delay_samples))
			stats["uart_delay_max_ms"] = 1000.0 * max(self.uart_delay_samples)

		return stats

File: head_tracking_audio/tools/matlab/latency_monitor.m
classdef LatencyMonitor < handle
    properties
        window = 200;
        samples = [];
        queue_age = [];
        uart_delay = [];
        _start = [];
    end
    methods
        function obj = LatencyMonitor(window)
            if nargin>0
                obj.window = window;
            end
            obj.samples = zeros(0,2);
            obj.queue_age = [];
            obj.uart_delay = [];
        end

        function start_callback(obj)
            obj._start = tic;
        end

        function end_callback(obj, frames, sample_rate)
            if isempty(obj._start)
                return
            end
            elapsed = toc(obj._start);
            budget = double(frames)/double(sample_rate);
            obj.samples(end+1,:) = [elapsed, budget];
            if size(obj.samples,1) > obj.window
                obj.samples(1,:) = [];
            end
            obj._start = [];
        end

        function record_queue_age(obj, age_sec)
            if isempty(age_sec), return; end
            obj.queue_age(end+1) = age_sec;
            if numel(obj.queue_age) > obj.window
                obj.queue_age(1) = [];
            end
        end

        function record_uart_delay(obj, delay_sec)
            if isempty(delay_sec), return; end
            obj.uart_delay(end+1) = delay_sec;
            if numel(obj.uart_delay) > obj.window
                obj.uart_delay(1) = [];
            end
        end

        function s = stats(obj)
            if isempty(obj.samples) && isempty(obj.queue_age) && isempty(obj.uart_delay)
                s = [];
                return;
            end
            s = struct();
            if ~isempty(obj.samples)
                elapsed = obj.samples(:,1);
                budget = obj.samples(:,2);
                s.avg_ms = 1000*mean(elapsed);
                s.max_ms = 1000*max(elapsed);
                s.overruns = sum(elapsed > budget & budget>0);
                s.budget_ms = 1000*mean(budget);
            end
            if ~isempty(obj.queue_age)
                s.queue_age_avg_ms = 1000*mean(obj.queue_age);
                s.queue_age_max_ms = 1000*max(obj.queue_age);
            end
            if ~isempty(obj.uart_delay)
                s.uart_delay_avg_ms = 1000*mean(obj.uart_delay);
                s.uart_delay_max_ms = 1000*max(obj.uart_delay);
            end
        end
    end
end
